src/mbgl/util/action_journal_impl.cpp

```
@@ -0,0 +1,522 @@
+        }
+    }
+
+    ~ActionJournalEvent() = default;
+
+    template <typename T>
+    ActionJournalEvent& addEvent(const rapidjson::GenericStringRef<char>& key, const T& value) {
+        eventJson.AddMember(key, value, json.GetAllocator());
+        return *this;
+    }
+
+    ActionJournalEvent& addEventStringArray(const rapidjson::GenericStringRef<char>& key,
+                                            const std::vector<std::string>& value) {
+        rapidjson::Value arrayJson(rapidjson::kArrayType);
+
+        for (const auto& elem : value) {
+            rapidjson::Value elemJson(rapidjson::kStringType);
+            elemJson.SetString(elem.c_str(), static_cast<rapidjson::SizeType>(elem.size()));
+            arrayJson.PushBack(elemJson, json.GetAllocator());
+        }
+
+        eventJson.AddMember(key, arrayJson, json.GetAllocator());
+        return *this;
+    }
+
+    template <typename T>
+    ActionJournalEvent& addEventEnum(const rapidjson::GenericStringRef<char>& key, const T& value) {
+        eventJson.AddMember(key, rapidjson::StringRef(Enum<T>::toString(value)), json.GetAllocator());
+        return *this;
+    }
+
+    std::string toString() {
+        if (!eventJson.ObjectEmpty()) {
+            json.AddMember("event", eventJson, json.GetAllocator());
+        }
+
+        rapidjson::StringBuffer jsonBuffer;
+        rapidjson::Writer<rapidjson::StringBuffer> jsonWriter(jsonBuffer);
+
+        json.Accept(jsonWriter);
+
+        return jsonBuffer.GetString();
+    }
+
+    rapidjson::Document json;
+    rapidjson::Value eventJson;
+};
+
+ActionJournal::Impl::Impl(const Map& map_, const ActionJournalOptions& options_)
+    : map(map_),
+      options(options_),
+      scheduler(Scheduler::GetSequenced()) {
+    assert(!options.path().empty());
+    assert(options.logFileSize() > 0);
+    assert(options.logFileCount() > 1);
+
+    options.withPath(options.path() + "/" + ACTION_JOURNAL_FOLDER_NAME);
+
+    openFile(detectFiles(), false);
```

Returns a bool, presumable to indicate success. But failure is silently ignored?

---

The open is retried on each subsequent, but indeed the initial error should be logged.